File: src/search/pipeline/keyword_extractor.py
"""
KIVO Search Engine
Keyword Extractor

Aufgabe: Query-Tokens gewichten. Bekannte "Konzepte" (von der Discovery
entdeckt) bekommen mehr Gewicht als Zufallswoerter.
"""

from __future__ import annotations


class KeywordExtractor:

    def __init__(self, analysis) -> None:
        self._analysis = analysis

    def execute(self, context) -> None:
        weighted = {}
        for term in context.normalized_tokens:
            weighted[term] = self._analysis.concept_weight(term)
        context.weighted_tokens = weighted

File: src-tauri/src/engine/pipeline/keyword_extractor.rs
//! KIVO Search Engine (Rust-Port)
//! Keyword Extractor
//!
//! 1:1-Port von `src/engine/search/pipeline/keyword_extractor.py`.
//! Aufgabe: Query-Tokens gewichten. Bekannte "Konzepte" (von der Discovery
//! entdeckt) bekommen mehr Gewicht als Zufallswoerter.

use std::collections::HashMap;

use crate::engine::analysis::Analysis;
use crate::engine::pipeline::context::PipelineContext;

pub struct KeywordExtractor<'a> {
    analysis: &'a Analysis,
}

impl<'a> KeywordExtractor<'a> {
    pub fn new(analysis: &'a Analysis) -> Self {
        Self { analysis }
    }

    /// Entspricht `KeywordExtractor.execute(context)`.
    pub fn execute(&self, context: &mut PipelineContext) {
        let mut weighted: HashMap<String, f64> = HashMap::new();
        for term in &context.normalized_tokens {
            weighted.insert(term.clone(), self.analysis.concept_weight(term));
        }
        context.weighted_tokens = weighted;
    }
}
